docs/source/features/parallelisms.rst

```diff
@@ -237,7 +237,7 @@ To utilize Sequence Parallelism in NeMo, set the ``sequence_parallel`` parameter
 
        sequence_parallel: True  # Enable Sequence Parallelism
 
-For further information on configuration, refer to the following documentation: `NeMo Megatron GPT Config <https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml#L66>`_.
+For further information on configuration, refer to the following documentation: `NeMo Megatron GPT Config <https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml#L66>`__.
```

Comment:
I am providing line-by-line copyedits to this file here because most of the file is read only.

Line 6. revise.
NeMo Megatron supports various data-parallel and model-parallel deep learning workload deployment methods, which can be mixed together arbitrarily.

Line 34. fix punctuation.
This approach significantly reduces the memory need of large-scale LLM training.

Line 41. change to NeMo Framework.
In NeMo Framework, DDP is the default parallel deployment method.

Line 42. fix punctuation.
This means that the total number of GPUs corresponds to the size of the DP group, and training an LLM with model parallelism decreases the size of the DP group.

Line 44. change to NeMo Framework.
Currently, NeMo Framework supports optimizer distribution only for Adam optimizer.

Lines 66-69. change to NeMo Framework.
DDP in NeMo Framework uses either PyTorch DistributedDataParallel (default) or a custom implementation if custom multi-precision training is enabled with megatron_amp_O2.

Line 71.change to NeMo Framework.
The distributed optimizer in NeMo Framework is built on top of
`DistributedFusedAdam 

Line 75. remove acronym (FSDP) in title. it is defined in the next sentence. This also makes it consistent with the other titles in this file.

Fully-Shared Data Parallelism

Line 78-81 revise.
NeMo Framework supports Fully-Sharded Data Parallelism (FSDP), which shards parameter gradients and low-precision parameters for computation. This is in addition to the model states that the distributed optimizer shards, including optimizer states and high-precision parameters.

Since FSDP shards the entire model states, it ensures linear model state memory savings with increasing DP size. FSDP is preferred for LLM training with unbalanced workloads between pipeline stages (or Transformer layers) or with a large vocabulary size, where pipelining would cause significant computation bubbles due to workload imbalance. 

Additionally, FSDP eliminates the need to search for performance-optimal mappings with 3D parallelism (TP/PP/DP) because it operates within a single parallelization domain.

Lines 83-86 revise.
NeMo Framework uses PyTorch’s FSDP interface to shard LLM model states, flattening the parameters of each transformer layer and partitioning them across data-parallel GPUs. FSDP introduces collective operations across data-parallel GPUs, including all-gather for parameter computation and reduce-scatter for parameter gradients. The all-gather operation occurs during both the network forward and back-propagation phases, while the gradient reduce-scatter operation happens only during back-propagation. These FSDP communications are overlapped with transformer layer computations.

Lines 96-97. change to NeMo Framework and fix capitalization.
Model Parallelism (MP) is a distributed model deployment method that partitions the model parameters across GPUs to reduce the need of per-GPU memory.
NeMo Framework supports various model-parallel methods, which can be mixed to maximize LLM training performance.

Lines 102-104. revise.
Tensor Parallelism (TP) is a model-parallel partitioning method that distributes the parameter tensor of an individual layer across GPUs. In addition to reducing model state memory usage, it also saves activation memory as the per-GPU tensor sizes shrink. However, the reduced per-GPU tensor size increases CPU overhead due to smaller per-GPU kernel workloads.

Line 114. fix capitalization.
To enable TP in the NeMo Framework, configure the ``tensor_model_parallel_size`` parameter in the model configuration. This parameter determines the number of GPUs among which the model's tensors are partitioned.

Line 116. delete **For Tensor Parallelism**: this bold text is not needed.

Line 129. change to NeMo Framework and use acronym.
NeMo Framework integrates TP through the implementation from Megatron Core. To understand how TP is activated within transformer blocks, refer to the code in the following repository: `Megatron-LM Transformer Block <https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/transformer_block.py>`_.

Lines 136-138 revise, use acronym, fix punctuation.
NeMo Framework supports FSDP along with TP. This is done by restricting the model state sharding to the data-parallel domain. Using FSDP with TP can be helpful when the model doesn't have sufficient parallelism to deploy on a large-scale training system with the data-parallel mapping. For example, running a model with the global batch size of 1024 on 2048 GPUs. Also, TP enables FSDP feasibility by reducing the model state size and the activation size per GPU, thus lower the FSDP communication overhead and the activation memory overhead.

Lines 140-141. revise.
Unset the CUDA_DEVICE_MAX_CONNECTIONS environment variable to set the number of GPU kernel queues, allowing the overlap of FSDP communication with computation kernels.

Line 157. revise.
To utilize Pipeline Parallelism (PP) in NeMo Framework, set the ``pipeline_model_parallel_size`` parameter in the model's configuration. This parameter specifies the number of GPUs among which the model's layers are distributed.

Line 159. delete **For Pipeline Parallelism**: this bold text is not needed.

Line 183. change to NeMo Framework.
The NeMo Framework implementation of PP leverages functionalities from Megatron Core. For a practical example of how PP is implemented within transformer blocks in NeMo, you can inspect the following codebase: `Megatron-LM Transformer Block <https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/transformer_block.py>`_.

Line 200. revise (we try to avoid "desired").
To enable EP, set ``model.expert_model_parallel_size`` to the expert parallel size you want. For example, if the model has six experts (``model.num_moe_experts=6``), then setting ``model.expert_model_parallel_size=3`` results in each GPU processing two experts. The number of experts should be divisible by the expert parallel size.

Line 212. change to NeMo Framework and use acronym.

The NeMo Framework implementation of EP uses functionality from Megatron Core. Please consult the `Megatron Core MoE layer <https://github.com/NVIDIA/Megatron-LM/blob/e2ec14ab5690fead7e33760b0f8fb20c83b4fd1f/megatron/core/transformer/moe/moe_layer.py#L29>`_ for more MoE implementation details.

Line 219. change to NeMo Framework.
NeMo Framework provides effective activation distribution methods, which is critical in training LLM with a large sequence length or large per-GPU micro-batch size.

Line 224. revise and add acronym.
Sequence Parallelism (SP) extends tensor-level model parallelism by distributing computing load and activation memory across multiple GPUs along the sequence dimension of transformer layers. This method is particularly useful for portions of the layer that have previously not been parallelized, enhancing overall model performance and efficiency.

Line 234. use acronym and change to NeMo Framework.
To utilize SP in NeMo Framework, set the ``sequence_parallel`` parameter to ``True`` in the model's configuration. Note that this feature is effective only when the tensor parallel size (``tensor_model_parallel_size``) is greater than ``1``.

Line 245.use acronym and change to NeMo Framework.
The NeMo Framework implementation of SP utilizes functionality from Megatron Core. For an in-depth look at how Sequence Parallelism is integrated into the Megatron Core architecture, you can examine the source code here: `Megatron-LM Sequence Parallel Source Code <https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/layers.py>`_.

Line 251. use acronym and revise with pronoun which.
Unlike SP, which partitions the activations of specific layers, CP divides the activations of all layers.

Line 256. fix capitalization.
To activate CP in the NeMo Framework, set the ``context_parallel_size`` parameter in the model configuration. This parameter specifies the number of GPUs among which the model's sequence activations are distributed.

Line 258. delete **For Context Parallelism**: it is not needed here.

Line 271. change to NeMo Framework.
NeMo Framework leverages functionalities from both Megatron Core and Transformer Engine to implement CP efficiently. During forward propagation, each GPU handles a segment of the sequence, storing only the necessary Key and Value (KV) pairs. In the backward pass, these KV pairs are reassembled across GPUs using advanced communication schemes like all-gather and reduce-scatter transformed into point-to-point communications in a ring topology. This method reduces the memory footprint significantly while maintaining computational efficiency.